syntax = "proto3";
package messaging;

/////////// Controller /////////////////////
message ErrorMsg {
    Vector3_msg position = 1; 
    Vector3_msg orientation = 2; 
    Vector3_msg vel = 3; 
    Vector3_msg angular_vel = 4; 
    Motor_msg motor_pos = 5; 
    Motor_msg motor_vel = 6;
    double time = 7; 
}

/////////// Motor Messages /////////////////
/// Command 
message MotorCmds_msg { 
    MotorCmd_msg FR_Hip = 1;
    MotorCmd_msg FR_Thigh = 2;
    MotorCmd_msg FR_Calf = 3;
    MotorCmd_msg FL_Hip = 4;
    MotorCmd_msg FL_Thigh = 5;
    MotorCmd_msg FL_Calf = 6;
    MotorCmd_msg RR_Hip = 7;
    MotorCmd_msg RR_Thigh = 8;
    MotorCmd_msg RR_Calf = 9;
    MotorCmd_msg RL_Hip = 10; 
    MotorCmd_msg RL_Thigh = 11; 
    MotorCmd_msg RL_Calf = 12;
    double time = 13; 
}

message MotorCmd_msg { 
    double Kp = 1; 
    double Kd = 2; 
    double pos = 3; 
    double vel = 4; 
    double tau = 5; 
}

/// Readings 
message MotorReadings_msg {
    Motor_msg torques = 1; 
    Motor_msg positions = 2; 
    Motor_msg velocities = 3; 
    double time = 5; 
}

message Motor_msg { 
    double FR_Hip = 1;
    double FR_Thigh = 2;
    double FR_Calf = 3;
    double FL_Hip = 4;
    double FL_Thigh = 5;
    double FL_Calf = 6;
    double RR_Hip = 7;
    double RR_Thigh = 8;
    double RR_Calf = 9;
    double RL_Hip = 10; 
    double RL_Thigh = 11; 
    double RL_Calf = 12;
    double time = 13;
}

////////////// Sensors/ EKF Messages /////////////////
message EKF_msg { 
    Vector3_msg position = 1; 
    Quaternion_msg quaternion = 2;
    Vector3_msg velocity = 3;
    Vector3_msg acceleration_bias = 4; 
    Vector3_msg angular_velocity_bias = 5;
    Vector3_msg angular_velocity = 6;
    double time = 7; 
}

message IMU_msg {
    Vector3_msg acceleration = 1;
    Vector3_msg gyroscope = 2;
    double time = 3;
}

message Vicon_msg { 
    Vector3_msg position = 1;
    Quaternion_msg quaternion = 2;
    double time = 3;
}

message Vector3_msg {
    double x = 1;
    double y = 2;
    double z = 3;
}

message Quaternion_msg {
    double w = 1;
    double x = 2;
    double y = 3;
    double z = 4;
}
